File: game/card.go
package game

import (
	"encoding/gob"
	"github.com/BigJk/end_of_eden/internal/lua/luhelp"
	"github.com/samber/lo"
	"strings"
)

func init() {
	gob.Register(CardInstance{})
}

// Card represents a playable card definition.
type Card struct {
	ID          string
	Name        string
	Description string
	Tags        []string
	State       luhelp.OwnedCallback
	Color       string
	PointCost   int
	MaxLevel    int
	DoesExhaust bool
	DoesConsume bool
	NeedTarget  bool
	Price       int
	Callbacks   map[string]luhelp.OwnedCallback
	Test        luhelp.OwnedCallback
	BaseGame    bool
}

func (c Card) PublicTags() []string {
	return lo.Filter(c.Tags, func(s string, i int) bool {
		return !strings.HasPrefix(s, "_")
	})
}

// CardInstance represents an instance of a card owned by some actor.
type CardInstance struct {
	TypeID string
	GUID   string
	Level  int
	Owner  string
}

func (c CardInstance) IsNone() bool {
	return len(c.GUID) == 0
}

File: assets/scripts/definitions/card.lua
---@meta

---@class card_state_ctx
---@field type_id type_id
---@field guid guid
---@field level number
---@field owner guid

---Card represents a playable card definition.
---@class card
---@field id? type_id
---@field name string
---@field description string
---@field tags? string[]
---@field state? fun(ctx:card_state_ctx):nil
---@field color string
---@field point_cost number
---@field max_level number
---@field does_exhaust? boolean
---@field need_target boolean
---@field price number
---@field callbacks callbacks
---@field test? fun():nil|string
---@field base_game? boolean

---CardInstance represents an instance of a card owned by some actor.
---@class card_instance
---@field guid guid
---@field type_id type_id
---@field level number
---@field owner guid